### ISSUE TITLE
Cordova iOS scheme "Modern (app:)" makes the screen go black upon refresh on iOS [was working for me in 368.2]

### ISSUE THREAD

**Issue by therealPaulPlay** [Scirra/Construct-bugs#7748]
## Problem description

When you are using the Cordova iOS scheme "modern" under Compatability settings, the iOS app will go blank (black) upon refresh when opening a different Webview Application (Opening a safari tab and going to a random web page works too!). Sometimes, it does so and almost instantly reloads the webview, but other times, it glitches out and you need to restart the app to make it load again (the latter is much more common).

~My old projects, that were updated to newer versions are, for some reason, not affected by this and will never refresh. Meanwhile, the new applications that use pretty much the same settings refresh like safari tabs (they become inactive after some time, then have to reload when opening them).~

My old projects are affected too when I update them from r368.2 to r375. There must have been a change in between these beta versions. Also, when I recreate the new project in r368.2, it works fine. Lastly, it's not necessarily a black screen, it's the background color of the application. 

## Attach a .c3p

[CordovaiOSSchemeBug.zip](https://github.com/Scirra/Construct-bugs/files/13967083/CordovaiOSSchemeBug.zip)

## Steps to reproduce

1. Creat a new project and fill in the details (description, name...)
2. Set/Keep the Cordova scheme on modern
3. I've tested this with Scale inner and outer
4. Export to iOS
5. Open the game, then open (a, a different construct app, b, a safari tab (any other browser should work too))
6. Reopen the game

## Observed result

The screen goes black. (I had to make the video super short to make it work on github)

https://github.com/Scirra/Construct-bugs/assets/80539587/e156397c-398d-4483-a3dd-01767e13f403

## Expected result

The screen should not go black. 

## More details
Broken in beta 375, working in r368.2
Did some testing: 370 is broken too, 368.2 works fine 100%

## System details

<details><summary>View details</summary>

PASTE HERE

</details>

**Comment by therealPaulPlay** [comment]
By testing this issue, I've also found out why you (probably) weren't able to replicatee #7686. The empty pod file (which results in an xcode error) only generates when the **Cordova iOS and android scheme** is set to **Legacy.**

**Comment by therealPaulPlay** [comment]
I've done further testing with this, as I've been wondering: Why has this not impacted my older projects?
Turns out, after deleting everything in an old project, changing most settings back to the default and adding (basically) the same image as in the provided test project, it still doesn't occur. Even more interesting, I'm getting this message in the Xcode console when repeating the steps that should replicate the bug: `CDVWebViewEngine shouldReloadWebView reload: 0`

I am not getting this message with the test project that I have provided earlier, as it does refresh (and go to a black screen). My old projects seem to be resistant to this. I don't know if that makes any sense. For good measure, I have tested this in r368.2 stable and yet, I'm getting the blackscreen in the new project and no issues in the old one. Huh. The old project (that is working) is provided here: 
[OldProject(noBlackScreen).zip](https://github.com/Scirra/Construct-bugs/files/13968638/OldProject.noBlackScreen.zip)

I'm no expert, but maybe you guys can work out the difference between these two projects to see what could cause this one to be "immune" against the black screen issue. It also uses the modern ios scheme (although It once used the legacy scheme).

**Comment by therealPaulPlay** [comment]
Edit: 
r374 is affected too (same as with r375, sometimes it only refreshes briefly and fades back in, sometimes it's stuck)
r372 is affected too with the same behavior
r370 is affected (only took around 20s to replicate the issue)
368.2 is not affected (I tested it thoroughly again, for 15 minutes)

**Comment by therealPaulPlay** [comment]
After reading the patchnotes, I think it‘s likely that it didn’t break in 370 but rather r369, maybe it has something to do with the added web worker support or the few ios bug fixes that happened in that update, maybe it’s also because of the HTML layers change, idk.. But in 368.2 construct apps don‘t refresh after inactivity like web pages in safari do (so basically, go blank and then fade in again, except the c3 projects often times don‘t fade in again) but exactly that happens in the new betas

**Comment by Tarek200** [comment]
I can confirm that when using ------ > (Cordova iOS Scheme = Modern) 

**On r368.2**
When the "App Refresh" the canvas shifts Up 

**On r369**
When the "App Refresh" it goes to a black screen


So the issue of the black screen starts from r369

**Comment by AshleyScirra** [comment]
I can't reliably reproduce this. It seemed to happen once or twice with the provided project as-is, so I did observe this happening, but not always. Being unable to reliably reproduce an issue makes it much harder to deal with - we might have to try 10-20 different code changes to fix a bug, and if each time it's not clear if it's fixed or if it's working by chance, it really slows things down and makes it much harder to tell what's happening. When I observed the issue happening, I could not find any error messages or anything unusual at all when using a debugger, so that did not provide any clues.

I tried adding a spinning sprite to the project. Then I could never reproduce the issue, no matter what I did. I tried putting the app in to the background, opening Safari, browsing some news website pages, opening a completely different Construct-made iOS app, going back to Safari, opening the other app again, then finally resuming the app in question; it never reproduced the blank screen even after repeatedly trying that. I also noticed that if all orientations are allowed, when you do get a blank screen, changing the orientation restores the correct display. The provided project is relatively unusual in that nothing changes and so Construct will not waste resources redrawing a screen that has not changed. So my best theory is that there is an iOS bug that causes a resumed app to show a blank screen until it next redraws. But this does not seem to explain the provided video which appears to be continually animated but still exhibits a blank screen on resume; but that project has not been provided so I don't know for sure, so I can't rule it out (this is why it's best to make sure issue reports refer solely to the provided project, and not other projects we can't test). The next release automatically redraws the screen when resuming an app in case that is the problem, so it would be worth testing that again, but I wouldn't be surprised if that's not it. But beyond that it's really hard to tell what's going wrong. Even though it appears to have only affected recent Construct releases, it could still be a bug in iOS; it's entirely possible that a system bug only affects a certain code path only in recent releases (e.g. worker mode) even though that code path is fully valid and correct. So this may still ultimately be something that has to be reported to Apple.

As it is difficult for me to reproduce the issue or identify what's going wrong, it would be useful if anyone affected could answer **all** of the following questions, as the more information that is available, the closer we can get to identifying what might be wrong. Please test repeatedly to be sure in your answers, as the issue seems intermittent, and answering questions based on chance results will mislead us and make it take longer to come to a solution.

1. Does setting 'Use worker' to 'No' avoid the problem? (Issue #7749 was related to worker mode which was activated on iOS in r369; it's possible that it caused this problem too, but for different reasons)
2. The original report mentions it's been tested with scale inner/outer fullscreen modes. Is "Letterbox scale" mode affected too? (This will help identify if the problem is specific to certain fullscreen modes)
3. If setting 'Use worker' to 'No' still reproduces the problem, and you have access to devices running an older version of iOS, e.g. iOS 16.x, does it reproduce there too? (If not this would imply a bug in iOS 17)
4. Does using a minimal project that constantly redraws (i.e. has a spinning sprite or some minimal amount of continuous animation) reproduce the problem?

**Comment by therealPaulPlay** [comment]
Thank you for looking into this. The provided video uses a different project because I was mainly testing with the one from the video and it does include animations etc. I just figured I‘d make the provided example project even more minimal, so I removed pretty much everything and could still reproduce the issue. However, I can definitely get the issue to replicate with projects in which there are a lot of moving objects etc. so maybe it depends on something else. I haven‘t tried it with a super minimal project that only features a spinning object though. 

What I have noticed however, is that this state (the inactivity / blackscreen) can also be triggered when construct has first loaded up (according to the console) but hasn‘t switched to the initial layout yet in my much bigger project. I assume this might be because this state takes significantly longer in the project that has ~40mb compressed. However, this makes this bug extremely severe for my use case as it literally breaks the game ~60% of the time (and I assume this could be the same for most big projects). 

I can reliably reproduce this issue by opening the safari browser, opening a new window and then loading a new web page that the browser hasn’t cached (so that it has to properly load it in). I‘ve noticed that if I visit the same website again, it sometimes doesn’t reproduce. Maybe my assumptions are wrong here, but I can at least reliably reproduce it with this method on my device.

I‘ll try to answer more of your questions later.

**Comment by therealPaulPlay** [comment]
2) Letterbox scale is indeed affected too. Also I've tested whether portrait mode makes a difference and it does not.

**Comment by therealPaulPlay** [comment]
3. Disabling the worker fixes the issue. When I'm opening a lot of new tabs etc. I could get it to reload, but it properly reloads without getting stuck on a black screen (or colored screen, depending on the background color of the app).

**Comment by therealPaulPlay** [comment]
4) I can't really replicate it in a project where there is e.g. an object with the sine behavior that is constantly changing in opacity. However, this is what happens during the loading screen of my bigger project, where a sprite is constantly tweening and using the wave effect:

**Comment by AshleyScirra** [comment]
One more question - does this happen with web exports in Safari as well?

I suppose if nothing else works and this appears to be an issue that only occurs in worker mode in iOS apps and not in Safari, we could just disable worker mode in iOS apps as a workaround.

**Comment by Tarek200** [comment]
Mm using:
r375
Worker Mode = Off
Scale Outer
A sprite rotating with Orbit behaviour

-I spent 4 hours yesterday and it did reload even once, I wasn't able to trigger the reload on this demo project when all the other C3 games installed reloaded like hundred times. Normally I could easily trigger the reload but not now using those especifications.

-However, after leaving it all night I opened it today again and it triggered the reload but it didn't go to a black screen so it just loaded normally and shifted the canvas up, though I observed something good when I closed the App by sending it to the background and I open it right away it fixed the Canvas offset without reloading the App, I will post this info on the canvas bug report see if it helps.

In short, I was able to reload the App just once after leaving it the whole night untouched and it didn't go to a black screen after reloading, but I still need to do more testing. Also, it looks like using those specifications and r375 it's been nearly impossible to trigger the reload when all the other C3 games reload nonstop.

This is the project that I used:

**Comment by AshleyScirra** [comment]
r376 is now out so it may be worth testing again with that in case it's any different.

**Comment by therealPaulPlay** [comment]
> One more question - does this happen with web exports in Safari as well?
> 
> I suppose if nothing else works and this appears to be an issue that only occurs in worker mode in iOS apps and not in Safari, we could just disable worker mode in iOS apps as a workaround.

I've tested it with r375 an hosted it here (https://modern-file-scheme-issue-ios.netlify.app) and came to the conclusion that it is affected too. 

It is a bit harder to replicate though, because you need to get the safari browser to do the refresh (I did so by opening like 20-30 other apps without closing them) which looks exactly like when construct games do this (it "fades" back in) and when you then switch back to the tab with the construct game, it is not loading and only displays a blackscreen in the case of this project. 

Even worse, when you reload the page, it only displays the construct loading screen, but once it has loaded back in, it goes straight to the blackscreen again. Restarting safari or reopening the tab when it's still in cache doesn't help either, however, changing the orientation of the device does work (as you said)

**Comment by therealPaulPlay** [comment]
> r376 is now out so it may be worth testing again with that in case it's any different. It now forces a redraw after resuming, which might help avoid a blank screen in some cases.

r376 seems to fix this - no blackscreen issues with the provided project anymore. Unfortunately, my main project still doesn't start / go to the initial layout in this version while it works with worker mode set to "Yes" in all browsers/exports (android webview/firefox/chrome). Maybe this is a different bug, I am not sure...

**Comment by therealPaulPlay** [comment]
So it turns out, while r376 fixes the blackscreen issue / unresponsiveness issue on refresh which is great, what actually still breaks my app, is that the [plugin](https://www.construct.net/en/make-games/addons/927/hide-homebutton-ios) that I'm using for handling proper fullscreen on iOS (= hiding the hombutton / making it semi transparent like in most games) suffers from the same issue, which means that it doesn't refresh after enabling the fullscreen mode in code. This only happens with versions above 368.2, likely because of the worker mode, but I'm not entirely sure. 

Maybe I could implement the same workaround that you have found in my plugin to also force a screen redraw, if you wouldn't mind sharing that solution and in case that would work for the plugin too (I'm not sure though how you are doing that). 

This is the simple project that I have been using for testing this (addon is bundled): 
[fullscreen-ios-test.zip](https://github.com/Scirra/Construct-bugs/files/14029069/fullscreen-ios-test.zip)

**Comment by AshleyScirra** [comment]
So as of r376 with worker mode disabled everything works correctly?

If so I think we will just disable worker mode in iOS apps. It's disabled in Android apps too for different reasons.

**Comment by therealPaulPlay** [comment]
Yeah, so tldr: r376 fixes the blackscreen/unresponsiveness bug on startup (for me at least) but worker mode still causes some issues on iOS. 

Additionally, I was able to replicate the bug where the canvas shifts up once in r376 but only on the construct loading screen and when the game had loaded up, it shifted back down - I‘m not sure whether that is related to the changes in r376 but that’s definitely a good thing and if that is indeed related, it might be a good idea to keep these changes around (though it doesn’t affect me directly as the plugin for the „immersive mode“ actually prevents the canvas shift bug from occurring)

**Comment by Tarek200** [comment]
I had limited access to the (Mac & iPhone with notch) so I was just able to replicate the reload by leaving the App closed for many hours then when I opened it triggered the reload and it loaded normally without any black screen, just the canvas brakes. I replicated this around 4 times.

However, I wasn't able to trigger the reload by swapping Apps, tabs, etc... I tried for hours with no luck yet. However, if the reload that triggers when the App is closed for long periods is the same as when it triggers when you swap between apps then I think the fix is working in r376 for the Black screen.  

> Additionally, I was able to replicate the bug where the canvas shifts up once in r376 but only on the construct loading screen and when the game had loaded up, it shifted back down - I‘m not sure whether that is related to the changes in r376 but that’s definitely a good thing and if that is indeed related, it might be a good idea to keep these changes around

Yeah, the workaround that you did for the canvas seems to work great the only thing is that it works after you (close and open) the App again, so if you could find a way that the workaround gets applied immediately after it triggers the reload I think it will sort out the Canvas bug swell. I posted a video on the canvas bug report showing how the workaround gets applied only when I close the App and open it again. However, in my example, I had ("Loader Style" = Off) so maybe that is why "therealPaulPlay" breaks on the loading screen then on the main layout it applies the fix so it didn't have to close and Open the App or could be something else.

**Comment by therealPaulPlay** [comment]
Yeah, it would be awesome if there was a way to detect the canvas shift and then apply this fix, but this is definitely good news already. Just fyi I was using the default c3 loading screen for this test

**Comment by AshleyScirra** [comment]
For the next beta I've made worker mode 'Auto' default to off in iOS exports only. Worker mode 'Yes' will still use it on iOS exports, but I would advise using the default, as providing an option to still use worker mode is mainly for diagnostic/testing purposes. As it happens Android exports also default 'Auto' to off for different reasons, but this means Android and iOS use the same settings. This does not affect web exports which will still use worker mode on both Android and iOS.

I suspect - but haven't been able to prove - the canvas offset is because the iOS webview is randomly scrolling the page. Nothing in Construct tries to scroll the page and nor does any scrolling happen on any other platform. So it would be an iOS bug if it suddenly scrolls the page for no particular reason. The code to reset the scroll position is this:

```js
document.scrollingElement.scrollTop = 0;
document.scrollingElement.scrollLeft = 0;
```

You can try using your own JavaScript code in Construct to use that to reset the scroll position if the canvas appears offset. I don't know if it'll work though as it's just a theory. I also added that code to reset the scroll position on startup in the next beta in case it helps with the loading screen. It's difficult to automatically apply that further, as when the virtual keyboard appears, the page is intentionally scrolled so the input control still remains in the visible part of the page; in that case the scroll position should not be reset as it could then scroll the input control off the view and so prevent you seeing what is being typed in. So Construct can automatically reset the scroll position in a few choice situations like startup, resuming the app etc. when the VK is not showing, but it can't just do it constantly.

If the scroll position is the cause of the offset bug, it is really an iOS bug and needs to be reported to Apple. I would not be surprised if there was such a bug in iOS as we've already found and worked around some other weird scroll position bugs that happen when the virtual keyboard appears/disappears. I'd have reported this issue to Apple myself already if I could reliably reproduce it, but it doesn't seem to happen often for me for some reason, so it would be best if someone who can more reliably reproduce it reports it to Apple directly. Adding code to reset the scroll position is a hack which really is the very definition of papering over cracks, and this type of thing can easily break in later releases, so the only truly reliable thing to do is have the real problem fixed by Apple.

So I think as of the next release we've done everything we can on our side - for any further issues, please go to Apple, as I really have no idea what is going on and it's proven difficult to make any meaningful changes on our side.

**Comment by therealPaulPlay** [comment]
Thank you very much for these tweaks and insights. Disabling the worker mode for iOS is probably a sensible solution to avoid any further issues including the blackscreen bug and I think, with the scroll reset now happening on the launch screen as well, the canvas shift isn't a huge problem anymore. 

Maybe @Tarek200 you can find a reliable solution to reproduce this bug, ideally just with a simple html canvas, but I unfortunately couldn't find a way to do that.

**Comment by Tarek200** [comment]
> Maybe @Tarek200 you can find a reliable solution to reproduce this bug, ideally just with a simple html canvas, but I unfortunately couldn't find a way to do that.

From Version r375 and up it has been impossible for me to trigger the reload by switching between Apps, I didn't manage to trigger it even once, while I could easily trigger it on 368.2. But the steps I gave to reproduce it in 368.2 didn't work for Ashley so I guess everyone has a different way to reproduce it. If I ever find a way to reliable reproduce it I will share it.
